src/librustc_target/spec/thumbv4t_nintendo_gba.rs

```diff
@@ -0,0 +1,80 @@
+//! Targets the Nintendo Game Boy Advance (GBA),
+//! a handheld game device from 2001.
+//!
+//! Please ping @Lokathor if changes are needed.
+//!
+//! The target profile assumes that you have the ARM binutils in your path (specifically the linker, `arm-none-eabi-ld`). They can be obtained for free for all major OSes from the ARM developer's website, and they may also be available in your system's package manager
+//!
+//! **Important:** This target profile **does not** specify a linker script or the ROM header. You'll still need to provide these yourself to construct a final  binary. Generally you'd do this with something like `-Clink-arg=-Tmy_script.ld` and `-Clink-arg=my_crt.o`.
+
+use crate::spec::{LinkerFlavor, PanicStrategy, RelocModel, Target, TargetOptions, TargetResult};
+
+pub fn target() -> TargetResult {
+    Ok(Target {
+        llvm_target: "thumbv4t-none-eabi".to_string(),
+        target_endian: "little".to_string(),
+        target_pointer_width: "32".to_string(),
+        target_c_int_width: "32".to_string(),
+        target_os: "none".to_string(),
+        target_env: "gba".to_string(),
+        target_vendor: "nintendo".to_string(),
+        arch: "arm".to_string(),
+        /* Data layout args are '-' separated:
+         * little endian
+         * stack is 64-bit aligned (EABI)
+         * pointers are 32-bit
+         * i64 must be 64-bit aligned (EABI)
+         * mangle names with ELF style
+         * native integers are 32-bit
+         * All other elements are default
+         */
+        data_layout: "e-S64-p:32:32-i64:64-m:e-n32".to_string(),
+        linker_flavor: LinkerFlavor::Ld,
+        options: TargetOptions {
+            linker: Some("arm-none-eabi-ld".to_string()),
+            linker_is_gnu: true,
```

Comment:
Other no_std thumb targets default to LLD, does that work for the GBA?

Reply:
I have to admit, I've never tried. The standard ARM binutils are the GNU binutils, so I've always just used `ld`. Also, most of the rest of the GBA community (outside of Rust I mean) uses a toolchain called "devkitproARM", which is also based on GNU and uses `ld`.

How would I go about testing lld? I don't even know where to get lld for arm-non-eabi

Reply:
I think you can just pass `-Clink-arg=-fuse-ld=lld` to rustc and it should work – LLD is shipped with Rust.

Reply:
Well I tested the one project I had sitting at hand, it worked, but that's a fairly small project which is nearly entirely in `main`.

I also now recall some of what @ketsuban told me when he first explained GBA stuff to me, that `lld` can't be used because it doesn't support `ARMv4T`. Based on [this PR](https://reviews.llvm.org/D53444) in the LLVM repo it seems like even recently they don't have support for ARMv4T in `lld`, and seemingly they don't plan to support it either.

The difficulty is that ARMv4T has a `BX` ("branch-exchange") instruction which is used to switch between a32 and t32 states in the CPU. Then in ARMv5T they added `BLX` ("branch-link-exchange"), which also leaves the next address in the link register as the branch-exchange happens (so that you can `bx lr` later to return). In some sitautions, LLD will use the BLX instruction, but ARMv4T doesn't support it, so you UB the CPU with an illegal instruction.

Reply:
But I don't have any experience with speaking to the LLVM folks, so maybe someone who does could ask them to formally add LLD support for ARMv4T in the future.

Reply:
I'll note that the extent of my experience here is having tried out LLD and noticed it emits a warning about using BLX on an unsupported architecture. I don't know under what circumstances it uses it, nor how I'd go about finding out.

Reply:
Okay, seems like we'll be using GNU LD for now in that case.